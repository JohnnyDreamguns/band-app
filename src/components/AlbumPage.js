import React from 'react';
import { Link } from 'react-router-dom';

const AlbumPage = ({ loading, album, band }) => (
  <div>
    {loading && <div className="loader"></div>}

    {album && band && (
      <div>
        <h2>{album.name}</h2>
        <p>Released: {album.released}</p>

        <p>
          Back to <Link to={`/band/${band.id}`}>{band.name}</Link>
        </p>
      </div>
    )}
  </div>
);

export default AlbumPage;
