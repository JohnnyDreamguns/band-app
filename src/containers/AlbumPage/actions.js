import {
  GET_ALBUM_BY_ID,
  SET_ALBUM_ID,
  CLEAR_ALBUM_PAGE,
  SET_LOADING
} from './constants';

export const getAlbumById = payload => ({
  type: GET_ALBUM_BY_ID,
  payload
});

export const setAlbumId = payload => ({
  type: SET_ALBUM_ID,
  payload
});

export const clearAlbumpage = () => ({
  type: CLEAR_ALBUM_PAGE
});

export const setLoading = payload => ({
  type: SET_LOADING,
  payload
});
