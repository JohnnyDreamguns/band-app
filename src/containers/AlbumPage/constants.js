export const GET_ALBUM_BY_ID = 'bands-app/AlbumPage/GET_ALBUM_BY_ID';
export const GET_BAND_BY_ID = 'bands-app/AlbumPage/GET_BAND_BY_ID';
export const SET_ALBUM_ID = 'bands-app/AlbumPage/SET_ALBUM_ID';
export const CLEAR_ALBUM_PAGE = 'bands-app/AlbumPage/CLEAR_ALBUM_PAGE';
export const SET_LOADING = 'bands-app/AlbumPage/SET_LOADING';
