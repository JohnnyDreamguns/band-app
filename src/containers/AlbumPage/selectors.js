import { initialState } from './reducer';
import { selectBandsData, selectAlbumsData } from '../../selectors/data';

export const albumRoot = state => state.albumPage || initialState;

export const selectAlbumId = state => albumRoot(state)['albumId'];

export const selectAlbum = state => {
  const albumId = selectAlbumId(state);
  if (!albumId) return undefined;
  return selectAlbumsData(state)[`${albumId}`];
};

export const selectLoading = state => albumRoot(state)['loading'];

export const selectBandId = state => {
  const album = selectAlbum(state);
  if (!album) return undefined;
  return album['band'];
};

export const selectBand = state => {
  const bandId = selectBandId(state);
  if (!bandId) return undefined;
  return selectBandsData(state)[`${bandId}`];
};
