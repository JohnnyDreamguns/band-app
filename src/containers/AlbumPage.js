import React from 'react';
import AlbumPage from '../components/AlbumPage';
import { useAlbumData } from '../hooks/useAlbumData';

const AlbumPageContainer = props => {
  const { album, band, loading, useFetchAlbumPageData } = useAlbumData();
  const { albumId } = props.match.params;

  useFetchAlbumPageData(albumId);

  const componentProps = {
    album,
    band,
    loading
  };

  return <AlbumPage {...componentProps} />;
};

export default AlbumPageContainer;
