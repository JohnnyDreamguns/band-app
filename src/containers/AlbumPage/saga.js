import { call, put, takeLatest, select } from 'redux-saga/effects';
import { GET_ALBUM_BY_ID } from './constants';
import { setLoading, setAlbumId } from './actions';
import { setError } from '../../actions/messages';
import { setBands, setAlbums } from '../../actions/data';
import { selectBandsData } from '../../selectors/data';

import request from '../../utils/request';

export function* getAlbumById(action) {
  const albumsURL = `http://localhost:3001/albums`;

  try {
    yield put(setLoading(true));

    const albumId = action.payload;

    const album = yield call(request, `${albumsURL}/${albumId}`);
    yield put(setAlbums(album.results));
    yield put(setAlbumId(parseInt(Object.keys(album.results)[0])));

    const bandId = album.results[albumId].band;
    const bandsData = yield select(selectBandsData);

    if (!bandsData.hasOwnProperty(bandId)) {
      yield getBandById(bandId);
    }

    yield put(setLoading(false));
  } catch (err) {
    yield put(setLoading(false));
    yield put(setError('There was a problem loading the album data'));
  }
}

export function* getBandById(bandId) {
  const bandsURL = `http://localhost:3001/bands`;

  const band = yield call(request, `${bandsURL}/${bandId}`);
  yield put(setBands(band.results));
}

export default function* bandpageDataSaga() {
  yield takeLatest(GET_ALBUM_BY_ID, getAlbumById);
}
