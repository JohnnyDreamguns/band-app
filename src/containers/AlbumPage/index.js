import React, { useEffect } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { getAlbumById, clearAlbumpage } from './actions';
import { useInjectSaga } from '../../utils/injectSaga';
import { useInjectReducer } from '../../utils/injectReducer';
import { selectLoading, selectAlbum, selectBand } from './selectors';
import saga from './saga';
import reducer from './reducer';

const key = 'albumPage';

export const AlbumPage = ({
  albumId,
  loading,
  album,
  band,
  getAlbumById,
  clearAlbumpage
}) => {
  useInjectReducer({ key, reducer });
  useInjectSaga({ key, saga });

  useEffect(() => {
    getAlbumById(albumId);

    return () => {
      clearAlbumpage();
    };
  }, [getAlbumById, clearAlbumpage, albumId]);

  return (
    <div>
      {loading && <div className="loader"></div>}

      {album && band && (
        <div>
          <h2>{album.name}</h2>
          <p>Released: {album.released}</p>

          <p>
            Back to <Link to={`/band/${band.id}`}>{band.name}</Link>
          </p>
        </div>
      )}
    </div>
  );
};

export const mapStateToProps = (state, ownProps) => {
  const { albumId } = ownProps.match.params;

  return {
    albumId,
    loading: selectLoading(state),
    album: selectAlbum(state),
    band: selectBand(state)
  };
};

export const mapDispatchToProps = dispatch => ({
  getAlbumById: albumId => dispatch(getAlbumById(albumId)),
  clearAlbumpage: () => dispatch(clearAlbumpage())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AlbumPage);
