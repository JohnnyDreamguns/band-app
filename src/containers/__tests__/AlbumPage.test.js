import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import TestRenderer from 'react-test-renderer';
import AlbumPageContainer from '../AlbumPage';
import AlbumPage from '../../components/AlbumPage';

const mockUseFetchAlbumPageData = jest.fn();

jest.mock('../../hooks/useAlbumData', () => ({
  useAlbumData: () => ({
    album: 'mock album',
    albumId: 5,
    band: 'mock band',
    bandId: 2,
    loading: true,
    clearAlbumId: 'mock clearAlbumId',
    useFetchAlbumPageData: mockUseFetchAlbumPageData,
    fetchAlbumPageData: 'mock fetchAlbumPageData',
    selectFirstBand: 'mock selectFirstBand',
    setLoading: 'mock setLoading',
    setAlbumId: 'mock setAlbumId'
  })
}));

jest.mock('../../hooks/useGlobalData', () => ({
  useGlobalData: () => ({
    setGlobalAlbums: () => {},
    setGlobalBands: () => {}
  })
}));

describe('AlbumPage container', () => {
  const props = {
    match: { params: { albumId: 5 } }
  };

  const testRenderer = TestRenderer.create(
    <BrowserRouter>
      <AlbumPageContainer {...props} />
    </BrowserRouter>
  );
  const testInstance = testRenderer.root;

  it('should match the props sent to the component', () => {
    expect(testInstance.findByType(AlbumPage.type).props).toMatchSnapshot();
  });

  it('should call mockUseFetchAlbumPageData', () => {
    expect(mockUseFetchAlbumPageData).toHaveBeenCalledTimes(1);
  });
});
