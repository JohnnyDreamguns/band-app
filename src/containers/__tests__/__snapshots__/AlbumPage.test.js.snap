// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlbumPage container should match the props sent to the component 1`] = `
Object {
  "album": "mock album",
  "band": "mock band",
  "loading": true,
}
`;
